kms/asymmetric.go

```diff
@@ -0,0 +1,260 @@
+func encryptRSA(keyName string, plaintext []byte) ([]byte, error) {
+	ctx := context.Background()
+	client, err := cloudkms.NewKeyManagementClient(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// Retrieve the public key from KMS.
+	response, err := client.GetPublicKey(ctx, &kmspb.GetPublicKeyRequest{Name: keyName})
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch public key: %+v", err)
+	}
+	// Parse the key.
+	block, _ := pem.Decode([]byte(response.Pem))
+	abstractKey, err := x509.ParsePKIXPublicKey(block.Bytes)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse public key: %+v", err)
+	}
+	rsaKey := abstractKey.(*rsa.PublicKey)
+	// Encrypt data using the RSA public key.
+	ciphertext, err := rsa.EncryptOAEP(sha256.New(), rand.Reader, rsaKey, plaintext, nil)
+	if err != nil {
+		return nil, fmt.Errorf("encryption failed: %+v", err)
+	}
+	return ciphertext, nil
+}
+
+// [END kms_encrypt_rsa]
+
+// [START kms_sign_asymmetric]
+
+// signAsymmetric will sign a plaintext message using a saved asymmetric private key.
+// example keyName: "projects/PROJECT_ID/locations/global/keyRings/RING_ID/cryptoKeys/KEY_ID/cryptoKeyVersions/1""
+func signAsymmetric(keyName string, message []byte) ([]byte, error) {
+	// Note: some key algorithms will require a different hash function.
+	// For example, EC_SIGN_P384_SHA384 requires SHA-384.
+	ctx := context.Background()
+	client, err := cloudkms.NewKeyManagementClient(ctx)
+	if err != nil {
+		return nil, err
+	}
+	// Find the digest of the message
+	digest := sha256.New()
+	digest.Write(message)
+	// Build the signing request
+	req := &kmspb.AsymmetricSignRequest{
+		Name: keyName,
+		Digest: &kmspb.Digest{
+			Digest: &kmspb.Digest_Sha256{
+				Sha256: digest.Sum(nil),
+			},
+		},
+	}
+	// Call the API.
+	response, err := client.AsymmetricSign(ctx, req)
+	if err != nil {
+		return nil, fmt.Errorf("asymmetric sign request failed: %+v", err)
+	}
+	return response.Signature, nil
+}
+
+// [END kms_sign_asymmetric]
+
+// [START kms_verify_signature_rsa]
+
+// verifySignatureRSA will verify that an 'RSA_SIGN_PSS_2048_SHA256' signature is valid for a given message.
+// example keyName: "projects/PROJECT_ID/locations/global/keyRings/RING_ID/cryptoKeys/KEY_ID/cryptoKeyVersions/1""
+func verifySignatureRSA(keyName string, signature, message []byte) error {
+	ctx := context.Background()
+	client, err := cloudkms.NewKeyManagementClient(ctx)
+	if err != nil {
+		return err
+	}
+
+	// Retrieve the public key from KMS.
+	response, err := client.GetPublicKey(ctx, &kmspb.GetPublicKeyRequest{Name: keyName})
+	if err != nil {
+		return fmt.Errorf("failed to fetch public key: %+v", err)
+	}
+	// Parse the key.
+	block, _ := pem.Decode([]byte(response.Pem))
+	abstractKey, err := x509.ParsePKIXPublicKey(block.Bytes)
+	if err != nil {
+		return fmt.Errorf("failed to parse public key: %+v", err)
+	}
+	rsaKey := abstractKey.(*rsa.PublicKey)
```

Comment:
Use the "comma ok" pattern to check this assertion?
```suggestion
	rsaKey, ok := abstractKey.(*rsa.PublicKey)
	if !ok {
		return fmt.Errorf(...)
	}
```